### ISSUE TITLE
Differential bignum fuzzer -- soliciting feedback

### ISSUE THREAD

**Issue by guidovranken** [google/oss-fuzz#1137]
I've built a differential bignum fuzzer at https://github.com/guidovranken/bignum-fuzzer that performs mathematical operations (add, subtract, mul, ...) in sync across a number of bignum implementations (currently OpenSSL, mbed TLS, C++ Boost multiprecision, Golang and Rust) and aborts on a mismatch.

This is a non-standard fuzzing project because:
- it can employed to 1) detect memory bugs or 2) to detect invalid computations or 3) both at the same time.
- Any number of bignum implementations (modules) can be activated, but with each additional module, the overall fuzzing process becomes slower.
- it has extra command-line parameters that can be used to specify such things as bignum input size and negative numbers yes/no (some BN implementations perform modulus and right-shift on negative numbers different than others, and it is necessary to disable negative numbers when both implementations are active). These extra command-line parameters are implemented in the libFuzzer-specific ```LLVMFuzzerInitialize``` function, and hence will not work as expected with AFL.

I think these points warrant some discussion as to how the project shall be deployed. An overly feature-rich configuration (sanitizers, mem leak detections, large inputs, comparing results across 5 modules) will incur such a speed penalty that (certain) bugs (if any) will not be found, but the same holds true for a too strict configuration. Perhaps 2 or 3 distinct deployments, each with a different configuration, is possible?

I am welcoming suggestions before I submit the project formally.

**Comment by kcc** [comment]
> it can employed to 1) detect memory bugs or 2) to detect invalid computations or 3) both at the same time.

This doesn't make it unusual. We have several targets like this in e.g. boringssl and openssl.

> Golang and Rust

This makes it unusual, unless we can link everything into one binary 
Today, oss-fuzz does not allow DSOs or additional binaries for fuzz targets.

> Any number of bignum implementations (modules) can be activated, but with each additional module, the overall fuzzing process becomes slower.

That's fine, it's totally worth it IMHO. 

> it has extra command-line parameters

For OSS-Fuzz we should try to avoid those. 
Some LLVM fuzz targets encode the command line flags in the binary name, which is ok. 

> Perhaps 2 or 3 distinct deployments, each with a different configuration, is possible?

Just create a single project, and keep adding fuzz targets to it. As many as you see fit (use common sense) 


BTW, there were some interesting ideas about improving differential fuzzing in libfuzzer -- with your set of targets we may have good guinea pigs to experiment with. Cool stuff!

**Comment by kcc** [comment]
> Golang 

Also, Go doesn't have coverage instrumentation that works with libFuzzer. (AFIACT). 
You will either have to build Go code w/o coverage, making such fuzzers less useful,
or use LLVM-GO, but I don't know if it's in usable state today. 
Rust should be just fine. 

I'd suggest you to start small and then iterate.

**Comment by guidovranken** [comment]
Thank you Kostya. I tried using LLVM-GO but it didn't seem very complete. Instead, for Go fuzzing I've been using a customized version of https://github.com/dvyukov/go-fuzz with my own version of libFuzzer that supports "custom guidance" ( https://github.com/guidovranken/libfuzzer-gv ). This works very well. But I digress.

I've made a PR with a single target that fuzzes OpenSSL master (instrumented with system-supplied CFLAGS) against Go (uninstrumented). I think that despite Go being uninstrumented this is a useful endeavour because 1. this way OpenSSL has a baseline to compare against (by comparing against Go, it finds several recent CVE's in OpenSSL's x86_64 assembly code, for instance) 2) Go bugs can still be found (I found a Go modexp bug recently without any instrumentation).

I'm planning to also add the latest OpenSSL 1.0.2 and 1.1.0 releases, as they may contain bugs that are not in master. My approach to linking everything together is such that there won't be any object name collisions during linking. So I could construct a fuzzer that compares 1.0.2, 1.1.0 and master against each other. Does this sound like a good idea to you?
